File: src/main/java/leetcode/hot100/LeetCode78.java
package leetcode.hot100;

import java.util.ArrayList;
import java.util.List;

/**
 * @author aiwujingxin@gmail.com
 * @date 2022/9/11 19:08
 */
public class LeetCode78 {

    public List<List<Integer>> subsets(int[] nums) {

        if (nums == null || nums.length == 0) {
            return new ArrayList<>();
        }
        List<List<Integer>> res = new ArrayList<>();

        dfs(nums, 0, res, new ArrayList<>());

        return res;
    }

    private void dfs(int[] nums, int index, List<List<Integer>> res, ArrayList<Integer> temp) {
        res.add(new ArrayList<>(temp));

        for (int i = index; i < nums.length; i++) {
            temp.add(nums[i]);
            dfs(nums, i + 1, res, temp);
            temp.remove(temp.size() - 1);
        }
    }
}

File: src/main/go/leetcode/lists/hot100/LeetCode78.go
package hot100

func subsets(nums []int) [][]int {
	var res [][]int
	var dfs func([]int, int, []int, *[][]int)
	dfs = func(nums []int, index int, temp []int, res *[][]int) {
		if index > len(nums)-1 {
			return
		}
		*res = append(*res, append([]int(nil), temp...))
		for i := index; i < len(nums); i++ {
			temp = append(temp, nums[i])
			dfs(nums, i+1, temp, res)
			temp = temp[0 : len(temp)-1]
		}
	}
	dfs(nums, 0, []int{}, &res)
	return res
}
